common/common.go

```diff
@@ -328,20 +335,53 @@ func SystemdVersion(systemdBinaryPath string) (int, error) {
 }
 
 // FSSupportsOverlay checks whether the filesystem under which
-// a specified path resides is compatible with OverlayFS
-func FSSupportsOverlay(path string) bool {
+// a specified path resides is compatible with OverlayFS.
+//
+// It returns an instance of ErrOverlayUnsupported if OverlayFS is not supported
+// or any other error if determining overlay support failed.
+func FSSupportsOverlay(path string) error {
 	var data syscall.Statfs_t
 	err := syscall.Statfs(path, &data)
 	if err != nil {
-		return false
+		return err
 	}
 
 	if data.Type == FsMagicAUFS ||
 		data.Type == FsMagicZFS {
-		return false
+		return ErrOverlayUnsupported("unsupported filesystem")
 	}
 
-	return true
+	dir, err := os.OpenFile(path, syscall.O_RDONLY|syscall.O_DIRECTORY, 0755)
+	if err != nil {
+		return errwrap.Wrap(fmt.Errorf("cannot open %q", dir), err)
+	}
+	defer dir.Close()
+
+	buf := make([]byte, 4096)
+	// ReadDirent forwards to the raw syscall getdents(3),
+	// passing the buffer size.
+	n, err := syscall.ReadDirent(int(dir.Fd()), buf)
+	if err != nil {
+		return errwrap.Wrap(fmt.Errorf("cannot read directory %q", dir), err)
+	}
+
+	offset := 0
+	for offset < n {
+		// offset overflow cannot happen, because Reclen
+		// is being maintained by getdents(3), considering the buffer size.
+		dirent := (*syscall.Dirent)(unsafe.Pointer(&buf[offset]))
+		offset += int(dirent.Reclen)
+
+		if dirent.Ino == 0 { // File absent in directory.
+			continue
+		}
+
+		if dirent.Type == syscall.DT_UNKNOWN {
+			return ErrOverlayUnsupported("unknown file type")
```

Comment:
Uhm, this looks a bit confusing in the resulting output. What about "unsupported filesystem, missing d_type support"?